air-quality-backend/src/air_quality/etl/forecast/forecast_data.py

```diff
@@ -80,18 +110,77 @@ def __init__(self, single_level_data: xr.Dataset, multi_level_data: xr.Dataset):
     _cached_pressure = None
     _cached_temperature = None
 
-    def _get_data_set(self, pollutant_type: PollutantType) -> xr.Dataset:
-        if is_single_level(pollutant_type):
+    def _get_data_set(self, forecast_data_type: ForecastDataType) -> xr.Dataset:
+        if is_single_level(forecast_data_type):
             return self._single_level_data
         else:
             return self._multi_level_data
 
+    def enrich_in_situ_measurements(
+            self,
+            in_situ_measurements: list[InSituMeasurement],
+            required_data: list[ForecastDataType]
+    ) -> list[tuple[InSituMeasurement, dict[ForecastDataType: float]]]:
+        if len(in_situ_measurements) == 0:
+            return []
+
+        latitudes = []
+        longitudes = []
+        valid_times = []
+        for in_situ_measurement in in_situ_measurements:
+            longitude = in_situ_measurement["location"]["coordinates"][0]
+            latitude = in_situ_measurement["location"]["coordinates"][1]
+            valid_time = in_situ_measurement["measurement_date"].timestamp()
+
+            if latitude not in latitudes:
+                latitudes.append(latitude)
+            if longitude not in longitudes:
+                longitudes.append(longitude)
+            if valid_time not in valid_times:
+                valid_times.append(valid_time)
+
+        interpolated_data_by_data_type = {}
+        for required_datum in required_data:
+            dataset = (self
+                       ._get_data_set(required_datum)
+                       .swap_dims({"step": "valid_time"}))
+
+            interpolated_data = dataset[required_datum.value].interp(
+                latitude=xr.DataArray(latitudes, dims="latitude"),
+                longitude=xr.DataArray(longitudes, dims="longitude"),
+                valid_time=xr.DataArray(valid_times, dims="valid_time"),
+                method="linear",
+            )
+            (interpolated_data
+             .set_index(valid_time=["valid_time"])
+             .set_index(latitude=["latitude"])
+             .set_index(longitude=["longitude"]))
```

Comment:
What does this actually do? Is it like a database index?

Reply:
I think so, the interp creates a dataset without indicies, and without index's the .sel call just fails. I think it's an index to help with searching, and due to xarray being tuned for large datasets it refuses to search on anything unindexed (a bit of a guess)